src/katgpucbf/xbgpu/bsend.py

```diff
@@ -0,0 +1,390 @@
+
+        self._chunks_queue: asyncio.Queue[Chunk] = asyncio.Queue()
+        buffers: list[np.ndarray] = []
+
+        # `n_heaps_to_send` is actually used to dictate the amount of buffers (in XSend)
+        # So perhaps we need to change the number of buffers to be range(send_free_queue.maxsize)
+        # n_heaps_to_send = len(buffers) // spectra_per_heap
+
+        send_shape = (heaps_per_fengine_per_chunk, self.n_beams, n_channels_per_substream, spectra_per_heap, COMPLEX)
+        for _ in range(n_tx_items):
+            chunk = Chunk(
+                accel.HostArray(send_shape, SEND_DTYPE, context=context),
+                accel.HostArray(
+                    (heaps_per_fengine_per_chunk, self.n_beams),
+                    np.uint32,
+                ),
+                channel_offset=channel_offset,
+                timestamp_step=timestamp_step,
+            )
+            self._chunks_queue.put_nowait(chunk)
+            buffers.append(chunk.data)
+
+        # Multicast stream parameters
+        self.heap_payload_size_bytes = n_channels_per_substream * spectra_per_heap * COMPLEX * SEND_DTYPE.itemsize
+        # Transport-agnostic stream information
+        # Used in XSend to calculate `send_rate_bytes_per_second`, do we need it here?
+        # TODO: Scope to move this calculation into a helper in utils
+        # packets_per_heap = math.ceil(self.heap_payload_size_bytes / packet_payload)
+        # packet_header_overhead_bytes = packets_per_heap * BSend.header_size
+
+        stream_config = spead2.send.StreamConfig(
+            max_packet_size=packet_payload + BSend.header_size,
+            max_heaps=n_tx_items * heaps_per_fengine_per_chunk + 1,  # TODO: Update this to be proper
+            rate_method=spead2.send.RateMethod.AUTO,
+            rate=0.0,  # TODO: Update to use `send_rate_bytes_per_second`, this sends as fast as possible
+        )
+        self.stream = stream_factory(stream_config, buffers)
+
+        item_group = spead2.send.ItemGroup(flavour=FLAVOUR)
+        item_group.add_item(
+            FREQUENCY_ID,
+            "frequency",  # Misleading name, but it's what the ICD specifies
+            "Value of the first channel in collections stored here.",
+            shape=[],
+            format=IMMEDIATE_FORMAT,
+        )
+        item_group.add_item(
+            TIMESTAMP_ID,
+            "timestamp",
+            "Timestamp provided by the MeerKAT digitisers and scaled to the digitiser sampling rate.",
+            shape=[],
+            format=IMMEDIATE_FORMAT,
+        )
+        item_group.add_item(
+            BF_RAW_ID,
+            "bf_raw",
+            "Channelised complex data.",  # TODO: Straight from the ICD, is this sufficient?
+            shape=buffers[0].shape[2:],
+            dtype=buffers[0].dtype,
+        )
+
+        self.descriptor_heap = item_group.get_heap(descriptors="all", data="none")
+
+    def enable_substream(self, stream_id: int = 0, enable: bool = True) -> None:
```

Comment:
I'd remove the default for `substream_id`. Any code that calls it without passing the substream ID is going to need to be updated sooner or later, and it would be better if it was obviously passing a hard-coded stream ID.